### PR TITLE
Writing debugging page

### PR DESCRIPTION
**Pull request by SylLutze**
Page with some info on model debugging. Let me know if the example is unclear. I considered adding photos but didn't want to overcomplicate. If pics would be helpful though lemme know and I can try and force an error to add photos of a stack trace, etc. 

Also including the engineers here if they have other thoughts to include or edit.

### CODE REVIEW

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+Making Sense of the Stack Trace
+-------------------------------
+
+The stack trace is an output that the model provides to show what 
+error message was received and trace back the flow of code to what 
+caused the error. Often the immediate line of code that generated the error will be 
+significantly removed from the line of code that will need to be changed 
+to fix the error. 
+
+Let's work through an example to understand this. Say a line of code
```
Comment (aflaxman) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:49]: Consider including an example of code that results in such a stack trace, so the reader can see exactly what you are referring to.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+in the code. If the dataframe is still correct, the issue is later. 
+Either use up or step (:code:`s`) to move up or "step" down to the next 
+line. In this manner, you can navigate around the code in the debugger and 
+find which line is the issue. From there, you can investigate the line 
+and fix the bug. 
+
+If you need to use multiple breakpoints, you can also use the 
+command :code:`c` for continue, to move until the next breakpoint. 
+
+Using iPython
```
Comment (aflaxman) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:149]:
```suggestion
Using `ipython`
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+-------------
+
+iPython is less commonly used for the type of debugging outlined
```
Comment (aflaxman) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:152]:
```suggestion
`ipython` is less commonly used for the type of debugging outlined 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+Using iPython
+-------------
+
+iPython is less commonly used for the type of debugging outlined 
+above. Instead, it is helpful for doing some basic checks on 
+outputs, or manipulation of outputs without opening a Jupyter 
+notebook. 
+
+First, you'll need to install iPython in your environment with
```
Comment (aflaxman) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:157]: oh, is iPython the name for the install?  maybe my changes above are wrong... I thought I was always typing `ipython`, though.

**Review by aflaxman — APPROVED** [review]

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 4 above)
Comment (zmbc) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:157]: I think it's actually IPython with a capital "I" ([ref](https://en.wikipedia.org/wiki/IPython))

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+Basic Debugging
+===============
+
+.. contents::
+   :local:
+   :depth: 2
+
+Debugging is an art and science that we couldn't hope to cover on this 
+single page. Rather than attempt such foolishness, we we review a few simple 
+techniques in the context of a targeted example here. If you find the
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:36]:
```suggestion
techniques in the context of a targeted example here. If you find your
```

Maybe?

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+techniques in the context of a targeted example here. If you find the 
+bug to be outside the scope of this page, it's likely time to ask an 
+engineer for help.  
+
+Making Sense of the Stack Trace
+-------------------------------
+
+The stack trace is an output that the model provides to show what 
+error message was received and trace back the flow of code to what 
+caused the error. Often the immediate line of code that generated the error will be
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:45]:
```suggestion
When the model runs into an error, it outputs a "stack trace" or "traceback" along with the error message.
The stack trace shows what Python was doing when it encountered the error.
Often the immediate line of code that generated the error will be 
```

I feel like we should make this a little bit more general because a lot of times, the thing that caused the error isn't visible anywhere in the stack trace!

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+So the bottom line of the stack trace will likely say an error was thrown 
+in a pandas error message function that the lengths are not equal. From there, 
+the stack trace displays the code that lead here - for example the pandas 
+checking function that was run, then the pandas function that was called 
+(the multiplication), then the secondary 
+:code:`loader.py` function that tried to multiply with the empty dataframe. And still, 
+none of these will actually reveal the issue with filtering the dataset 
+we started with! This can be further complicated if the :code:`loader.py` function 
+using some functions for other packages, like :code:`vivarium_research` or :code:`vivarium_public_health`.
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:71]:
```suggestion
uses some functions for other packages, like :code:`vivarium_research` or :code:`vivarium_public_health`. 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+place to start looking for issues, and the stack trace will provide the line 
+number of the code that caused the error. 
+
+Otherwise, see the next section for how to use the python debugger. 
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the 
+flag :code:`--pbd` is used, which means that when an error
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:85]:
```suggestion
flag :code:`--pdb` is used, which means that when an error 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+you wrote or edited - like the :code:`loader.py` file. If so, this is a good 
+place to start looking for issues, and the stack trace will provide the line 
+number of the code that caused the error. 
+
+Otherwise, see the next section for how to use the python debugger. 
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:84]:
```suggestion
In all of the commands used to run Vivarium operations, the 
```

Maybe? Or just "in almost all Vivarium commands"?

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+number of the code that caused the error. 
+
+Otherwise, see the next section for how to use the python debugger. 
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the 
+flag :code:`--pbd` is used, which means that when an error 
+occurs, you will be placed into the python debugger.
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:86]: nit: capitalize Python throughout

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+===============
+Basic Debugging
+===============
+
+.. contents::
+   :local:
+   :depth: 2
+
+Debugging is an art and science that we couldn't hope to cover on this 
+single page. Rather than attempt such foolishness, we we review a few simple
```
Comment (pletale) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:35]:
```suggestion
single page. Rather than attempt such foolishness, we review a few simple 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+flag :code:`--pbd` is used, which means that when an error 
+occurs, you will be placed into the python debugger. 
+
+The python debugger is nice to use in tandem with the stack trace. 
+Let's continue with the example above, where we got an error 
+in a pandas checking function. 
+
+The python debugger places you at the end of the stack trace - 
+so we are currently sitting in a pandas error message function. 
+We can check this by typing :code:`l.` or :code:`ll` which prints
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:94]: What does `l.` do? I always just use `l`.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+Otherwise, see the next section for how to use the python debugger. 
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the 
+flag :code:`--pbd` is used, which means that when an error 
+occurs, you will be placed into the python debugger. 
+
+The python debugger is nice to use in tandem with the stack trace.
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:88]: I don't really consider this a stack trace -- suggest removing this line.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+Let's work through an example to understand this. Say a line of code 
+we edit in the :code:`loader.py` file filters a dataset incorrectly 
+such that it is actually empty. Nothing in that immediate block of 
+code catches this error - instead the provided dataframe is just 
+stored as empty. 
+
+From there this same, empty dataset is used in another :code:`loader.py` 
+function to calculate a rate. So our erroreously empty dataframe will be 
+multiplied by a dataframe with the correct data in it. 
+When pandas tries to by this multiplication on the dataframes,
```
Comment (pletale) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:58]:
```suggestion
When pandas tries to run this multiplication on the dataframes, 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+The python debugger places you at the end of the stack trace - 
+so we are currently sitting in a pandas error message function. 
+We can check this by typing :code:`l.` or :code:`ll` which prints 
+the code around the current line (l.), or a longer version of this 
+print out (ll). This will let you see what the code surrounding the 
+error is. Likely in this case, it will just show you that pandas found 
+mismatched lengths and threw the error message, which isn't too helpful. 
+
+Next, you can move up or down the stack trace with :code:`u` or
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:100]: I think when you say "stack trace" here, it should be "call stack." The "trace" is just how that stack gets output in e.g. an error message. It could also be good to explain what the "stack" is: it's the code you're currently running, and what function called the current one, and what function called that one, and so on, all the way up the top.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+Maybe you see the error quickly. But if not, we'll probably 
+want to add a breakpoint. 
+
+To quit the python debugger, enter :code:`quit`. 
+
+Adding Breakpoints
+------------------
+
+So, we have identified the issue is an empty dataframe. Let's keep 
+debugging! The datframe in question is generated
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:129]:
```suggestion
debugging! The dataframe in question is generated 
```

Comment (pletale) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:129]:
```suggestion
debugging! The dataframe in question is generated 
```

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+A breakpoint is a line of code :code:`breakpoint()` that "breaks" 
+the code. So when this line is read, the model will drop you into a 
+python debugger at that point in the code. So in our example, we might 
+set a breakpoint in the function which creates the empty dataframe. 
+
+Once in the python debugger, we can use a similar process to the above. 
+Start by printing the dataframe. If it's empty, the issue is earlier 
+in the code. If the dataframe is still correct, the issue is later. 
+Either use up or step (:code:`s`) to move up or "step" down to the next 
+line. In this manner, you can navigate around the code in the debugger and
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:142]:
```suggestion
You can use up and down to see functions calling the current one, as in the debugger.
You can also step forward (:code:`s`) to run the next line of code and then drop back into
the debugger.
In this manner, you can check what your dataframes look like before and after each
line runs and
```

I think it's good to explain step in a bit more detail.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+Using iPython
+-------------
+
+iPython is less commonly used for the type of debugging outlined 
+above. Instead, it is helpful for doing some basic checks on 
+outputs, or manipulation of outputs without opening a Jupyter 
+notebook. 
+
+First, you'll need to install iPython in your environment with 
+:code:`pip install iPython`. Then start an iPython session by
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:158]:
```suggestion
:code:`pip install ipython`. Then start an iPython session by 
```

I think PyPI package names are canonically lowercase.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+code can't be stored and you must enter and run lines one at a time. Some effects 
+of this are, for example, you have to run all :code:`import` lines every 
+time you start an iPython session, and you can't store custom 
+functions to use repeatedly easily. However, there are some use cases. 
+
+For example, if you want to print out an artifact key to check the 
+columns included. Jupyter notebook can do this, but you will need 
+to open a notebook, connect online, create a file, print the file, 
+and at the end of that, you'll need to end the Jupyter session before you 
+can run anything else from the open terminal. This is a lot of work to visualize
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:171]: Note: it is possible to start Jupyter so it doesn't hog your terminal!

**Review by pletale — APPROVED** [review]
Looks great, just noticed a couple typos!

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+
+From here, you can use this the same as a Jupyter notebook, except 
+code can't be stored and you must enter and run lines one at a time. Some effects 
+of this are, for example, you have to run all :code:`import` lines every 
+time you start an iPython session, and you can't store custom 
+functions to use repeatedly easily. However, there are some use cases. 
+
+For example, if you want to print out an artifact key to check the 
+columns included. Jupyter notebook can do this, but you will need 
+to open a notebook, connect online, create a file, print the file,
```
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:169]:
```suggestion
to open a notebook, connect online, create a notebook, load and print the data in the artifact,
```

This was a little hard to parse for me -- is this right?

(there is also a "console" feature in Jupyter that doesn't require creating a notebook, FWIW)

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 1 above)
Comment (NathanielBlairStahn) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:49]: I agree, I think having an actual code example would make this easier to follow.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 2 above)
Comment (zmbc) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:149]: I have a strong feeling based on this section that you might prefer to use VSCode for this purpose -- happy to demo if you'd like!

**Review by zmbc — APPROVED** [review]

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+-------------------------------
+
+The stack trace is an output that the model provides to show what 
+error message was received and trace back the flow of code to what 
+caused the error. Often the immediate line of code that generated the error will be 
+significantly removed from the line of code that will need to be changed 
+to fix the error. 
+
+Let's work through an example to understand this. Say a line of code 
+we edit in the :code:`loader.py` file filters a dataset incorrectly
```
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:50]: Also, I had no idea what `loader.py` was (I still don't, really, since I've never had to build an Artifact), and I didn't know where to look for it. So I looked at the other pages in this section and eventually found it mentioned on the Artifact Building page, but that page also didn't make it clear where the function came from, so I had to poke around until I found it in the `vivarium_research_template` repo [here](https://github.com/ihmeuw/vivarium_research_template/blob/main/%7B%7Bcookiecutter.package_name%7D%7D/src/%7B%7Bcookiecutter.package_name%7D%7D/data/loader.py). So before jumping straight into the example, I think you should provide a bit of context, like "Suppose you're building an Artifact, as described on [[link to Artifact Buliding page]]," and when you mention the `loader.py` file, it would be helpful to link to the actual file so the reader can see what sorts of things are in there.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (NathanielBlairStahn) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:158]: Also, it would probably be good to mention that if you've already installed Jupyter, then IPython should already be installed, so you don't explicitly have to do this pip install.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+typing :code:`iPython` in the command line. 
+
+From here, you can use this the same as a Jupyter notebook, except 
+code can't be stored and you must enter and run lines one at a time. Some effects 
+of this are, for example, you have to run all :code:`import` lines every 
+time you start an iPython session, and you can't store custom 
+functions to use repeatedly easily. However, there are some use cases. 
+
+For example, if you want to print out an artifact key to check the 
+columns included. Jupyter notebook can do this, but you will need
```
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:168]: I'm not sure what this means (printing out an artifact key to check the columns included). Isn't the key just a string? Do you mean print out the data stored at that key? Could you give some example code?

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+time you start an iPython session, and you can't store custom 
+functions to use repeatedly easily. However, there are some use cases. 
+
+For example, if you want to print out an artifact key to check the 
+columns included. Jupyter notebook can do this, but you will need 
+to open a notebook, connect online, create a file, print the file, 
+and at the end of that, you'll need to end the Jupyter session before you 
+can run anything else from the open terminal. This is a lot of work to visualize 
+one dataframe! iPython can be run in the 
+terminal without creating unneeded files or worrying about connections.
```
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:173]: Are you assuming you have the Artifact stored locally on your laptop? If not, won't you still have to connect to the cluster to see the data?

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 19 above)
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:171]: You could always open multiple terminals, or use tmux to run multiple things simultaneously within a single terminal session.

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+columns included. Jupyter notebook can do this, but you will need 
+to open a notebook, connect online, create a file, print the file, 
+and at the end of that, you'll need to end the Jupyter session before you 
+can run anything else from the open terminal. This is a lot of work to visualize 
+one dataframe! iPython can be run in the 
+terminal without creating unneeded files or worrying about connections. 
+
+Therefore, while iPython is "worse" than Jupyter notebooks overall (in this 
+author's opinion), it is faster and therefore useful for quick tasks. 
+Common uses include: printing out dataframes saved elsewhere; simple
```
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:177]: What does "elsewhere" mean?

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 9 above)
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:84]: What do you mean by "engineering commands"? Can you give one or more examples of a command you might be running where you can use the `--pdb` flag?

File: docs/source/model_design/programming_vivarium/debugging/index.rst [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the 
+flag :code:`--pbd` is used, which means that when an error 
+occurs, you will be placed into the python debugger. 
+
+The python debugger is nice to use in tandem with the stack trace. 
+Let's continue with the example above, where we got an error
```
Comment (NathanielBlairStahn) [inline docs/source/model_design/programming_vivarium/debugging/index.rst:89]: Again, having the actual code in front of me would be helpful for following along with the example.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 4 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:157]: Fixed all instances of referencing IPython to have capital I and P and all code to be all lowercase ipython.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 9 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:84]: Changed a bit differently than this to be more clear.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 12 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:94]: Hmm interesting! The engineers just taught me `l.` - it prints out the code lines where you are. Not sure why they do this instead of just `l`. I can ask Jim

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 2 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:149]: Hey Zeb - maybe separately from this PR we can chat about IPython alternatives - the VSCode and having Jupyter not hog your terminal. I think I'm still going to merge this now but maybe next docs session we can run through these options and see what we want as the default in the docs.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 19 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:171]: Per other comment - adding this to be discussed later!

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 15 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:100]: Thanks!

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 19 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:171]: Going to add this to the convo with Zeb on options for these quick changes. Not changing in this PR but noted.

File: docs/source/model_design/programming_vivarium/debugging/index.rst (hunk 21 above)
Comment (SylLutze) [reply docs/source/model_design/programming_vivarium/debugging/index.rst:50]: Sure - I also linked to the overview of engineering files page which is designed to help explain the different files in the engineering repos. https://vivarium-research.readthedocs.io/en/latest/model_design/programming_vivarium/overview_engineering_files/index.html

**Comment by SylLutze** [comment]
Thanks all for the comments. I am planning to merge this since there are already lots of comments, but noting that there are a few outstanding items I'll include in a separate PR: 
- Example codes/photos 
- Possible changes in the IPython section for other alternative methods (to be discussed at the next docs session) 
- The `l.` vs `l` comment from Zeb that I wanted to follow up with engineering on.
